src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java

```diff
@@ -694,74 +692,82 @@ private HttpClientRequest createEnrichInventoryClientRequest(HttpClient httpClie
-              return;
-            }
-            logger.warn("Trying to request SRS again, attempts left: {}", attemptsCount.decrementAndGet());
-            vertx.setTimer(REREQUEST_SRS_DELAY, timer -> doPostRequestToSrs(srsClient, vertx, deletedRecordSupport, listOfIds, attemptsCount, promise));
-            return;
-          } else {
-            String errorMsg = getErrorFromStorageMessage("source-record-storage", srsResponse.request()
-              .absoluteURI(), srsResponse.statusMessage());
-            srsClient.close();
-            handleException(promise, new IllegalStateException(errorMsg));
-            return;
-          }
+          String errorMsg = getErrorFromStorageMessage("source-record-storage", srsResponse.request()
+            .absoluteURI(), srsResponse.statusMessage());
+          srsClient.close();
+          handleException(promise, new IllegalStateException(errorMsg));
+          return;
         }
-        srsResponse.bodyHandler(getSrsResponseBodyHandler(srsClient, promise));
-      });
+        srsResponse.bodyHandler(buffer -> handleSrsResponse(srsClient, promise, buffer));
+      }, e->  retrySRSRequest(srsClient, vertx, deletedRecordSupport, listOfIds, attemptsCount, retryAttempts, promise, 400, "Error in SRS response"));
     } catch (Exception e) {
       handleException(promise, e);
     }
   }
 
-  private Handler<Buffer> getSrsResponseBodyHandler(SourceStorageSourceRecordsClient srsClient, Promise<Map<String, JsonObject>> promise) {
-    return buffer -> {
-      final Map<String, JsonObject> result = Maps.newHashMap();
-      try {
-        final Object jsonResponse = buffer.toJson();
-        if (jsonResponse instanceof JsonObject) {
-          JsonObject entries = (JsonObject) jsonResponse;
-          final JsonArray records = entries.getJsonArray("sourceRecords");
-          records.stream()
-            .filter(Objects::nonNull)
-            .map(JsonObject.class::cast)
-            .forEach(jo -> result.put(jo.getJsonObject("externalIdsHolder")
-              .getString(INSTANCE_ID_FIELD_NAME), jo));
-        } else {
-          logger.debug("Can't process response from SRS: {}", buffer.toString());
-        }
-        promise.complete(result);
-      } catch (DecodeException ex) {
-        String msg = "Invalid json has been returned from SRS, cannot parse response to json.";
-        handleException(promise, new IllegalStateException(msg, ex));
-      } catch (Exception e) {
-        handleException(promise, e);
-      } finally {
-        srsClient.close();
+  private void retrySRSRequest(SourceStorageSourceRecordsClient srsClient, Vertx vertx, boolean deletedRecordSupport, List<String> listOfIds, AtomicInteger attemptsCount, int retryAttempts, Promise<Map<String, JsonObject>> promise, int statusCode, String statusMessage) {
+    String warnMessage = "Got error response form SRS, status code: " + statusCode + ", status message: " + statusMessage;
+    logger.debug(warnMessage);
```

Comment:
It's probably better to log this before calling this method.  The way this currently stands, in some cases you might log a message at ERROR level, then a different (also inaccurate - see earlier comments) message at WARN level.

Reply:
I still see you passing a hard-coded 400 into this method in the case where there is no response code - because the error occurred while writing the request.  Changing the log level from WARN to DEBUG is not sufficient to fix this IMO.

Reply:
> I still see you passing a hard-coded 400 into this method in the case where there is no response code - because the error occurred while writing the request. Changing the log level from WARN to DEBUG is not sufficient to fix this IMO.
@craigmcnally, it's not correct. 
There re-request can be triggered by two reasons: 
1) when the srs returns a full response but with incorrect status code
2) when there a problem in srs with passing the body and the timeout exception is thrown (or any else potential http exceptions).
These two cases are different situations and they should have different log messages. I corrected it and we didn't pass the hardcoded "400" value. 
![image](https://user-images.githubusercontent.com/60663044/110809886-f6f23480-828d-11eb-9c6f-a7988d100290.png)